File: Leetcode/110-Balanced-Binary-Tree/BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getHeight(self, node):
        if node == None:
            return 0
        left = self.getHeight(node.left)
        right = self.getHeight(node.right)
        if left == -1 or right == -1:
            return -1
        if abs(left-right)>1:
            return -1
        return max(left, right) + 1
    
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root == None:
            return True
        if self.getHeight(root) == -1:
            return False
        return True

File: Leetcode/110-Balanced-Binary-Tree/BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        if (getHeight(root)==-1) {
            return false;
        }
        return true;
    }
    
private:
    int getHeight(TreeNode* node) {
        if (node == NULL) {
            return 0;
        }
        int left = getHeight(node->left);
        int right = getHeight(node->right);
        if (left == -1 || right == -1) {
            return -1;
        }
        if (abs(left-right)>1) {
            return -1;
        }
        return max(left, right) + 1;
    }
};


